source/kernel/src/comms/rosc.rs

```diff
@@ -0,0 +1,185 @@
+//! Reusable One-Shot Channel
+//!
+//! Often, clients of drivers only want to process one "in-flight" message at
+//! a time. If request pipelining is not required, then a Reusable One-Shot Channel
+//! is an easy way to perform an async/await request/response cycle.
+//!
+//! Essentially, a Rosc is a single producer, single consumer, channel, with a max
+//! depth of one. Many producers can be created over the lifecycle of a single consumer,
+//! however only zero or one producers can be live at any given time.
+
+use core::{
+    cell::UnsafeCell,
+    mem::MaybeUninit,
+    sync::atomic::{AtomicU8, Ordering},
+};
+
+use maitake::wait::WaitCell;
+use mnemos_alloc::containers::HeapArc;
+
+use crate::Kernel;
+
+/// A reusable One-Shot channel.
+///
+/// A `Rosc<T>` can be used to hand out single-use [Sender] items, which can
+/// be used to make a single reply.
+///
+/// A given `Rosc<T>` can only ever have zero or one `Sender<T>`s live at any
+/// given time, and a response can be received through a call to [Rosc::receive].
+pub struct Rosc<T> {
```

Comment:
style nit, take it or leave it: i personally don't love that this is an acronym; if i see `Rosc` for the first time in other code, i'm going to have to go look at the declaration before i know what the hell a `Rosc` is...